djangocms_picture/models.py

```diff
@@ -1,83 +1,284 @@
+        self.picture = oldinstance.picture
+
+    def get_size(self, context, placeholder):
+        crop = self.use_crop or False
+        upscale = self.use_upscale or False
+        # use field thumbnail settings
+        if self.use_thumbnail:
+            width = self.use_thumbnail.width
+            height = self.use_thumbnail.height
+            crop = self.use_thumbnail.crop
+            upscale = self.use_thumbnail.upscale
+        elif self.use_automatic_scaling:
+            width = context.get('width', None)
+            height = context.get('height', None)
+        else:
+            width = self.width or None
+            height = self.height or None
+
+        # calculate height when not given according to the
+        # golden ratio
+        if not height and width:
+            height = int(width / PICTURE_RATIO)
+        if not width and height:
+            width = int(height * PICTURE_RATIO)
+
+        options = {
+            'size': (width, height),
+            'crop': crop,
+            'upscale': upscale,
+        }
+        return options
+
+    def get_link(self):
+        if self.link_url:
+            return self.link_url
+        if self.link_page:
+            return self.link_page.get_absolute_url(language=self.language)
+        return False
 
     def clean(self):
-        if self.url and self.page_link:
+        # there can be only one link type
+        if self.link_url and self.link_page:
             raise ValidationError(
-                _("You can enter a Link or a Page, but not both."))
+                ugettext('You have defined an external and internal link. '
+                    'Only one option is allowed.')
+            )
+
+        # you shall only set one image kind
+        if not self.picture and not self.external_picture:
+            raise ValidationError('You need to add either an image or '
+                'an external link to an image.')
+
+        # certain cropping options do not work together, the following
+        # list defines the disallowed options used in the ``clean`` method
+        invalid_option_pairs = [
+            ('use_automatic_scaling', 'use_no_cropping'),
+            ('use_automatic_scaling', 'use_thumbnail'),
+            ('use_no_cropping', 'use_crop'),
+            ('use_no_cropping', 'use_upscale'),
+            ('use_no_cropping', 'use_thumbnail'),
+            ('use_thumbnail', 'use_crop'),
+            ('use_thumbnail', 'use_upscale'),
+        ]
+        # invalid_option_pairs
+        invalid_option_pair = None
+
+        for pair in invalid_option_pairs:
+            if getattr(self, pair[0]) and getattr(self, pair[1]):
+                invalid_option_pair = pair
+                break
+
+        if invalid_option_pair:
+            field_1 = self._meta.get_field(invalid_option_pair[0])
+            field_2 = self._meta.get_field(invalid_option_pair[1])
+            message = ugettext('The cropping selection is not valid. '
```

Comment:
make sure the `.format()` call is **after** the `ugettext()`
and don't add spaces before or after the = sign here